pravega/pravega-operator
# Issue 78: Clean up persistent volumes when deleting Pravega Cluster

## Change log description

- Clean up persistent volumes when Pravega Cluster is deleted.

- Add ownerference to pvc. When Pravega Cluster is deleted, k8s will delete those pvc obejcts. If pv is not associated with a pvc, it will also be deleted by k8s automatically. 

## Purpose of the change
Fix [#78](https://github.com/pravega/pravega-operator/issues/78)

## How to test the change
Checkout to source branch, follow the instructions on the README.md to deploy a Pravega Cluster. 

After you have a running Pravega Cluster, check the pvc list
```
$ kubectl get pvc
cache-pravega-pravega-segmentstore-0   Bound    pvc-131c0c4c-f784-11e8-a645-42010a8000a3   20Gi       RWO            standard       21s
cache-pravega-pravega-segmentstore-1   Bound    pvc-131f464a-f784-11e8-a645-42010a8000a3   20Gi       RWO            standard       21s
cache-pravega-pravega-segmentstore-2   Bound    pvc-1323405c-f784-11e8-a645-42010a8000a3   20Gi       RWO            standard       21s
data-example-0                         Bound    pvc-d792c268-f75f-11e8-a645-42010a8000a3   20Gi       RWO            standard       4h19m
data-example-1                         Bound    pvc-f2876a1f-f75f-11e8-a645-42010a8000a3   20Gi       RWO            standard       4h18m
data-example-2                         Bound    pvc-0f791f0e-f760-11e8-a645-42010a8000a3   20Gi       RWO            standard       4h18m
journal-pravega-bookie-0               Bound    pvc-12f6fc86-f784-11e8-a645-42010a8000a3   10Gi       RWO            standard       21s
journal-pravega-bookie-1               Bound    pvc-13006daa-f784-11e8-a645-42010a8000a3   10Gi       RWO            standard       21s
journal-pravega-bookie-2               Bound    pvc-130dfc2b-f784-11e8-a645-42010a8000a3   10Gi       RWO            standard       21s
ledger-pravega-bookie-0                Bound    pvc-12f87cc8-f784-11e8-a645-42010a8000a3   10Gi       RWO            standard       21s
ledger-pravega-bookie-1                Bound    pvc-1305b35c-f784-11e8-a645-42010a8000a3   10Gi       RWO            standard       21s
ledger-pravega-bookie-2                Bound    pvc-130f8b08-f784-11e8-a645-42010a8000a3   10Gi       RWO            standard       21s
pravega-tier2                          Bound    pvc-32b64f5b-f4e1-11e8-a645-42010a8000a3   50Gi       RWX            nfs            3d8h
```

Now run the command to delete the Pravega Cluster. 
```
$ kubectl delete -f pravega.yaml
```

You can now  see the related pvc are deleted 
```
$ kubectl get pvc
NAME             STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-example-0   Bound    pvc-d792c268-f75f-11e8-a645-42010a8000a3   20Gi       RWO            standard       4h20m
data-example-1   Bound    pvc-f2876a1f-f75f-11e8-a645-42010a8000a3   20Gi       RWO            standard       4h19m
data-example-2   Bound    pvc-0f791f0e-f760-11e8-a645-42010a8000a3   20Gi       RWO            standard       4h19m
pravega-tier2    Bound    pvc-32b64f5b-f4e1-11e8-a645-42010a8000a3   50Gi       RWX            nfs            3d8h
```

And persistent volumes
```
$ kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS     CLAIM                    STORAGECLASS   REASON   AGE
pvc-0f791f0e-f760-11e8-a645-42010a8000a3   20Gi       RWO            Delete           Bound      default/data-example-2   standard                4h24m
pvc-32b64f5b-f4e1-11e8-a645-42010a8000a3   50Gi       RWX            Delete           Bound      default/pravega-tier2    nfs                     3d8h
pvc-56119aad-f365-11e8-bddf-42010a800161   50Gi       RWX            Delete           Released   default/pravega-tier2    nfs                     5d5h
pvc-d792c268-f75f-11e8-a645-42010a8000a3   20Gi       RWO            Delete           Bound      default/data-example-0   standard                4h26m
pvc-f2876a1f-f75f-11e8-a645-42010a8000a3   20Gi       RWO            Delete           Bound      default/data-example-1   standard                4h25m
```
Signed-off-by: wenqi <Wenqi.Mou@emc.com>
